### ISSUE TITLE
Problem parsing magnitude that starts with "0"

### ISSUE THREAD

**Issue by cpascual** [hgrecco/pint#614]
Hi, I am not sure if this is a bug, but at least it is quite unexpected to me:

```python
print Quantity('123').magnitude  # --> 123
print Quantity('0123').magnitude  # --> 0  
```

**Comment by cpascual** [comment]
If I get confirmation that this is unwanted behaviour, I can try to submit a PR for changing it

**Comment by dalito** [comment]
I find the result unexpected, too. But what is the wanted behaviour? Should pint match the Python interpreter? That may be surprising for newcomers in case of Python 2.x:
```
Python 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:19:30) [MSC v.1500 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.                        
>>> 0123                                                                                      
83     
```                                                                                       
```
Python 3.6.4 (v3.6.4:d48eceb, Dec 19 2017, 06:54:40) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> 0123
  File "<stdin>", line 1
    0123
       ^
SyntaxError: invalid token
>>>
```

**Comment by cpascual** [comment]
Yes, I would expect pint to strip units and then treat the rest just as python's eval would. 

So, in python 2:

```python
Quantity(0123)    # returns <Quantity(83, 'dimensionless')> ... OK
Quantity(0o123)  # returns <Quantity(83, 'dimensionless')> ... OK
Quantity(0123.)   # returns  <Quantity(123.0, 'dimensionless')> ... OK
Quantity('0123')  # returns  <Quantity(0, 'dimensionless')>
                  # ... but should return:  <Quantity(83, 'dimensionless')>
Quantity('0o123')  # raises UndefinedUnitError: 'o123m' is not defined in the unit registry.
                  # ... but should return:  <Quantity(83, 'dimensionless')>
Quantity('0123m') # returns  <Quantity(0, 'meter')>
                  # ... but should return: <Quantity(83, 'meter')>
Quantity('0123.')   # returns  <Quantity(123.0, 'dimensionless')>  OK
```

And in Python 3:
```python
Quantity(0123)  # raises SyntaxError: invalid token ... OK
Quantity(0o123)  # returns <Quantity(83, 'dimensionless')>  ... OK
Quantity(0123.)   # returns  <Quantity(123.0, 'dimensionless')> ... OK
Quantity('0123')  # returns  <Quantity(0, 'dimensionless')>
                   # ... but should raise SyntaxError: invalid token   
Quantity('0o123')  # raises UndefinedUnitError: 'o123m' is not defined in the unit registry.
                   # ... but should return:  <Quantity(83, 'dimensionless')>
Quantity('0123m') # returns  <Quantity(0, 'meter')>  
                    # ... but should return: <Quantity(83, 'meter')>
Quantity('0123.')   # returns  <Quantity(123.0, 'dimensionless')>  ... OK
```

**Comment by cpascual** [comment]
The above is IMHO the clearest behaviour (in the sense that one acan refer to pthon's eval for the expected  output)

But, personally, I would not object if `Quantity('0123')` returned `<Quantity(123, 'dimensionless')>` in both python 2 and 3

**Comment by dalito** [comment]
What you propose, looks all good to me.

**Comment by hgrecco** [comment]
Good to find you both of you in one issue! I would like to invite you to have commit rights to this. I have been very busy lately (I am sure you notice, sorry) To keep pint in good shape we need new blood and new ideas. You have contributed with code, ideas and discussion to pint and it will be great if your would like to join

**Comment by cpascual** [comment]
@hgrecco: I am glad to join, but I want to be up-front (at the risk of sounding selfish): I cannot commit much time to maintaining another project so if I join it would be mostly to focus on fixing/implementing those features that affect my main project (http://github.com/taurus-org). If you are ok with that, I am in!

**Comment by dalito** [comment]
@hgrecco: I am also willing to join but will be very limited in the time that I can commit to pint. In the end three people with little time will still be better than just one.

**Comment by hgrecco** [comment]
Awesome. My idea is to find a group of around five people that have contributed regularly to pint and to provide access. I perfectly understand (and suffer my self) the time constrains and I agree that a collaborative effort. Pint is a project that service others and therefore it makes perfect sense that bugs will be discovered when integrated in other project and features will be designed with certain projects in mind. So .. let's go ahead.

**Comment by dopplershift** [comment]
@hgrecco I'm happy to help where I can as well. The core library I'm responsible heavily leverages pint, so it's very important that it keep progressing.

**Comment by hgrecco** [comment]
I am using [bors](https://bors.tech/) to merge PR in Pint. It is very nice as it provides a consistent and safe way to merge PR. You can read the docs [here](https://bors.tech/documentation/getting-started/)

**TL;DR** Just comment any PR that you want to merge with `bors r+` to tell bors that you have approved it. You can do it with multiple PRs and bors will merge each of them sequentially testing them against the new master after each merge.

**Comment by hgrecco** [comment]
Would you like to find a PR that you like / feel comfortable of merging and merging through bors so we can see that everything is fine?

**Comment by hgrecco** [comment]
sounds perfect. Go for it!

**Comment by andrewgsavage** [comment]
This appears to be fixed on current `master`. Testing the cases discussed in this thread:

```python
>>> import pint
>>> Quantity = pint.Quantity
>>> Quantity('123').magnitude
123
>>> Quantity('0123').magnitude
123
>>> Quantity('0123m')
Quantity(123, 'meter')
>>> Quantity('0123.')
Quantity(123.0, 'dimensionless')
>>> Quantity('-0123')
Quantity(-123, 'dimensionless')
```

All match the behavior settled on in the discussion above (strip leading zeros like a normal decimal literal, no silent truncation to 0). I couldn't pin down the exact commit that fixed it, but it likely came along with the parser rewrite since Python 2 support was dropped.

One minor loose end from the thread: `Quantity('0o123')` still raises `UndefinedUnitError` rather than parsing as octal 83. That was a secondary suggestion in the discussion rather than the reported bug though, and doesn't seem worth adding given pint doesn't otherwise infer numeric base from string prefixes.

Closing since the original bug is resolved. Please reopen if you disagree or hit a case that still misbehaves.
